documentation/specs/question.md

```diff
@@ -0,0 +1,84 @@
+}
+else
+{
+  TaskLoggingHelper.LogMessageWithCodeFromResources("ToolTask.NotUpToDate");
+}
+```
+
+
+### `ToolTask`
+If inheriting from ToolTask, your custom task can override `SkipTaskExecution()`.  When it returns `false`, TookTask will exit with error.
+
+## Shipping Tasks
+When `/question` switch is used, it will modify the shipping task with these behavior.  Note: this is still experimental and can change.
+
+`Exec`
+Doesn't have an up-to-date check.  It will always run.  Use Target Inputs and Outputs to skip the exec task.
+
+`Touch`
+Warns when a file is touched.  It is unclear if the file touched will participate in the build as it is a common practice to touch a file to signal external tool to run.  Use Target Inputs and Outputs to skip this task.
+
+`Copy`
+Errors out when any copy action occurs.
+
+`WriteLinesToFile`
+Error when WriteOnlyWhenDifferent is true.  This task could be used to append to a log file that isn't participating in the build itself.
+
+`Delete`
+Warn that a file still exists and is to be deleted.  It is unclear if the file is part of the build or it is deleting an accessary file outside of the build.
+
+`Move`
+No warning or errors.  This task is effectively a no-op as files could not exist anymore.
+
+`DownloadFile`
+Error if SkipUnchangedFiles is true.
+
+`GenerateResource` 
+Error if any files would be generated.
+
+`MakeDir`
+Error if folder doesn't exist.
+
+`RemoveDir`
+Error if folder still exist.
+
+`Unzip`
+Error if SkipUnchangedFiles is true.
+
+`ZipDirectory`
+Error if the destination zip file doesn't exists.
+
+## Common Up-To-Date Error
+- **Typographical error**. Spelling or incorrect path.  Check if the target inputs and outputs real files.
+- **Casing**. While MSBuild on Windows is case insenstive, some command line comparision and hashing tool are case senstive.  Paths and Project Refernces are a common source of casing issues.
+- **Cross Product**. Inputs and Outputs are sometimes used for Cross Product. Try to move all to `Outputs` parameter or `Returns` parameter instead of Inputs.
+- **Exec Task** are not Skipable, thus they should be wrapped with Target Inputs and Outputs or other systems.  For backwards compatibility, Question will not issue an error.
+- **FileWritten**.  The common clean system will remove files that aren't in the FileWritten itemgroup.  Sometimes task output won't be add to FileWritten itemgroup.
+- **Build, Build, Build**.  Sometimes, a 2nd build will expose up to date issue.  Verify building 3 times. For example, 1st Build, Question 2nd build, Run 2nd Build, and Question 3rd build.
+- **Double Checks**.  Since target and task could be incremental, if both are implemented, then it can lead task skipping but not the task.  For example, a Target has inputs A and outputs B.  If A is newer, than B, then the target will start.  If the task compares the content of A and B and deems nothing has changed, then B is not updated.  If such case, this leads to target rerunning.
```

Comment:
```suggestion
- **Double Checks**.  Since target and task could be incremental, if both employ incrementality checks, then it can lead to skipping the task but not the calling target.  For example, a Target has inputs A and outputs B.  If A is newer, than B, then the target will start.  If the task called by that target compares the content of A and B and deems nothing has changed, then B is not updated.  In such case, it will lead to the target rerunning.
```

Btw. is there any advice to this case, how to avoid the situation? E.g. - would it ba advicable to use a temprary file with hash of the real input as the input to the target?

Reply:
It will differ case by case, but in general, to consolidate the checks into a single location and remove the other.  The author would need to decide what is the best course of action.